vllm/model_executor/models/whisper.py

```diff
@@ -0,0 +1,689 @@
+
+
+def dummy_encoder_data_for_whisper(ctx: InputContext, seq_len: int,
+                                   mm_counts: Mapping[str, int]):
+    assert mm_counts["audio"] == 1
+    sample_rate = 16000
+    num_tokens = ctx.model_config.hf_config.max_source_positions
+    return DummyData(
+        SequenceData.from_prompt_token_counts((0, num_tokens)),
+        {"audio": [(np.zeros(30 * sample_rate), sample_rate)]},
+    )
+
+
+@lru_cache
+def get_whisper_processor(
+    processor_name: str,
+    *args,
+    trust_remote_code: bool = False,
+    revision: Optional[str] = None,
+    **kwargs,
+) -> WhisperProcessor:
+    """Gets an whisper processor for the given model name via HuggingFace."""
+    return WhisperProcessor.from_pretrained(
+        processor_name,
+        *args,
+        trust_remote_code=trust_remote_code,
+        revision=revision,
+        **kwargs,
+    )
+
+
+def input_processor_for_whisper(ctx: InputContext, inputs):
+    multi_modal_data = inputs["encoder"]["multi_modal_data"]
+    if isinstance(multi_modal_data["audio"], list):
+        assert len(multi_modal_data["audio"]) == 1
+        multi_modal_data["audio"] = multi_modal_data["audio"][0]
+    # Resample and process audio
+    audio, orig_sr = multi_modal_data["audio"]
+    processor = get_whisper_processor(ctx.model_config.model)
+    target_sr = processor.feature_extractor.sampling_rate
+    audio = librosa.resample(audio, orig_sr=orig_sr, target_sr=target_sr)
+    if audio.size > 30 * target_sr:
+        # Truncate audio to 30 seconds
+        audio = audio[:30 * target_sr]
+    multi_modal_data["audio"] = (audio, target_sr)
+    # Calculate number of tokens after convolutions
+    num_tokens = (audio.size // 80 - 1) // 2 + 1
+    num_tokens = (num_tokens - 2) // 2 + 1
+    # Pre-allocate placeholder tokens in encoder sequence
+    inputs["encoder"]["prompt_token_ids"] = [0] * num_tokens
+    return inputs
+
+
+def input_mapper_for_whisper(
+    ctx: InputContext,
+    multi_modal_data: Union[np.ndarray, List[np.ndarray]],
+) -> MultiModalKwargs:
+    if not isinstance(multi_modal_data, list):
+        multi_modal_data = [multi_modal_data]
+
+    assert len(multi_modal_data) == 1
+
+    if len(multi_modal_data) == 0:
+        return MultiModalKwargs()
+
+    processor = get_whisper_processor(ctx.model_config.model)
+    sampling_rate = processor.feature_extractor.sampling_rate
+
+    audios = [audio for audio, _ in multi_modal_data]
+
+    kwargs = processor(audios, sampling_rate=sampling_rate,
+                       padding=False, return_tensors="pt")
+    kwargs["input_features"] = kwargs["input_features"].squeeze(0)
+    kwargs["input_features"] = kwargs["input_features"].to(torch.float16)
+
+    return MultiModalKwargs(kwargs)
+
+
+@INPUT_REGISTRY.register_dummy_encoder_data(dummy_encoder_data_for_whisper)
+@INPUT_REGISTRY.register_input_processor(input_processor_for_whisper)
+@MULTIMODAL_REGISTRY.register_input_mapper("audio", input_mapper_for_whisper)
```

Comment:
If possible, use the merged multi-modal processor (#10114) which should simplify your code.

Reply:
What's the right approach here? I tried the Processor and there appears to be several mismatched assumptions with Whisper and encoder-decoder models:

1. There is a mandatory `_get_prompt_replacements` which does not apply to Whisper (audio tokens never get embedded within text tokens).
2. The processor assumes one token sequence but for Whisper, two token sequences are needed: one as a placeholder for the encoder to allocate the correct cross-attention kv-cache, and the second for the actual prompt which only goes to the decoder.

(1) is easy to get around with some awkwardness, but (2) seems only supported by the legacy multimodal processing pipeline which returns two different inputs, one for the encoder and another for the decoder.

Reply:
Thanks for bringing this up! I haven't thought about encoder-decoder models for the new pipeline yet... let's use the legacy processing pipeline for now then.